cogs/file_cog.py

```diff
@@ -0,0 +1,141 @@
+
+
+async def save_jpg_with_target_size(im, filename, target):
+    # https://stackoverflow.com/questions/52259476/how-to-reduce-a-jpeg-size-to-a-desired-size/52281257#52281257
+    """Save the image as JPEG with the given name at best quality that makes less than "target" bytes"""
+    # Min and Max quality
+    Qmin, Qmax = 25, 96
+    # Highest acceptable quality found
+    Qacc = -1
+    while Qmin <= Qmax:
+        m = math.floor((Qmin + Qmax) / 2)
+
+        # Encode into memory and get size
+        buffer = io.BytesIO()
+        im.save(buffer, format="JPEG", quality=m)
+        s = buffer.getbuffer().nbytes
+
+        if s <= target:
+            Qacc = m
+            Qmin = m + 1
+        elif s > target:
+            Qmax = m - 1
+
+    # Write to disk at the defined quality
+    if Qacc > -1:
+        im.save(filename, format="JPEG", quality=Qacc)
+        filename.seek(0)
+        return filename
+    else:
+        print("ERROR: No acceptable quality factor found", file=sys.stderr)
+        return None
+
+
+class FileCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.Cog.listener()
+    async def on_message(self, message):
+        if message.author == self.bot.user or len(message.attachments) == 0:
+            return
+        added_pdf = False
+        for attachment in message.attachments:
+            if not added_pdf and any(
+                attachment.filename.endswith(i) for i in [".pdf", ".PDF"]
+            ):
+                yimin_emote = self.bot.get_emoji(733516365243220038)
+                await message.add_reaction(yimin_emote)
+                added_pdf = True
+            elif any(attachment.filename.endswith(i) for i in [".heic", ".HEIC"]):
+                async with aiohttp.ClientSession() as session:
+                    async with session.get(attachment.url) as r:
+                        file = io.BytesIO(await r.read())
+
+                heif_file = pyheif.read(file)
+                image = Image.frombytes(
+                    heif_file.mode,
+                    heif_file.size,
+                    heif_file.data,
+                    "raw",
+                    heif_file.mode,
+                    heif_file.stride,
+                )
+                with io.BytesIO() as image_binary:
+                    image_binary = await save_jpg_with_target_size(
+                        image, image_binary, 8e6
+                    )
+                    if image_binary is None:
+                        await message.channel.send("`ERROR: failed to resize image`")
+
+                    await message.channel.send(
+                        file=discord.File(
+                            fp=image_binary, filename=attachment.filename[:-5] + ".png"
+                        )
+                    )
+        return
+
+    @commands.Cog.listener()
+    async def on_raw_reaction_add(self, payload):
```

Comment:
I know I'm the one responsible for it, but we need better comments in this function since it's kinda complicated.